### ISSUE TITLE
Problems with `share_memory` on `Lazy` modules

### ISSUE THREAD

**Issue by belldandyxtq** [pfnet/pytorch-pfn-extras#47]
**Problem Statement**
Since the parameters in `Lazy` modules, like [LazyLinear](https://github.com/pfnet/pytorch-pfn-extras/blob/3ae7d279b5a20282265a64812a99c5bfbe18ecc1/pytorch_pfn_extras/nn/modules/lazy_linear.py#L7), changes on the first forward. If the model is shared with [share_memory](https://github.com/pytorch/pytorch/blob/df252c059c3ea4c04f8c39c3b4c25a3f4572b0b7/torch/nn/modules/module.py#L1327). The model may not be correctly shared as the new parameters will not be on the shared memory.

**Examples**
```
1. shared_model = some_lazy_model()
2. shared_model.share_memory()
3. fork to process A and B
4. process A: shared_model(input)... # forward and step
5. process B: test_model.load_state_dict(shared_model.state_dict()) # problems here!!!
```
The parameters changed in (4) are not on the shared memory which happened in (2), so (5) can not get updated results.

The problem was first noticed and reported by @shu65

**Comment by belldandyxtq** [comment]
We have decided to first generate error in such case #46

**Comment by emcastillo** [comment]
The problem here is that this requires to do always a forward pass in the main process before fork.
Otherwise memory won't be allocated and shared later.
Currently the forward pass needs to be done before calling `share_memory`.

**Comment by kmaehashi** [comment]
Closing as error message implemented in #46 and the upstream (https://github.com/pytorch/pytorch/pull/40493)
